### PR TITLE
[RFC] Implement SearchCurrent for current match hl

### LINKED ISSUE #4600: feature request: hl-IncSearch / hl-Search alternating during normal searches
With `hl-IncSearch` and `hl-Search` highlight groups set to alternate colors, during a search & replace, Neovim helpfully highlights the current match to replace with `hl-IncSearch`, as in this image:

hl-IncSearch

This feature request is in regards to normal searching, where the default behavior is to not alternate between `hl-IncSearch` and `hl-Search`. I think the default should be customizable, so that users can get consistent behavior with normal searching and search & replace.

So far I've found [SearchParty](https://github.com/dahu/SearchParty) provides this feature, but the plugin is quite large.

If it is better to not adjust behavior of hl-IncSearch, then how about an `hl-CurSearch` so that users can declare this as the wanted behavior?

### PR DESCRIPTION
**Pull request by certik**
Uses SearchCurrent to highlight the current match under cursor.

How to use: put this into your `init.vim`:
```
hi Search ctermbg=blue ctermfg=red
hi SearchCurrent ctermbg=green ctermfg=red
```
Then the current match will be highlighted using `SearchCurrent`, while the rest will be highlighted using `Search`. An example:

nvim_search_current

Issues to resolve:

* [x] When "n" is pressed, the colors do not get updated. One has to press `/int<CR>` (in the screenshot above) when the cursor is at some other place, and it will correctly highlight that new match. One has to figure how to redraw when "n" or "N" is pressed.
* [ ] Update tests for the new behavior (a few currently fail)
* [ ] document the new option `SearchCurrent`

Fixes #4600.

### CODE REVIEW

**Comment by blueyed** [comment]
For reference: related Vim issue: https://github.com/vim/vim/issues/2819

**Comment by justinmk** [comment]
would it make sense to re-use `IncSearch` for this instead of adding another highlight group?

**Comment by certik** [comment]
@justinmk I reused IncSearch, I think that's better. I now redraw the screen in `nv_next()`, and the problem is fixed! Both "n" and "N" now works. Thanks for the suggestion. Here is how it looks like, with default nvim settings:

nvim_search_current2

And here is how it looks like with `hi IncSearch ctermfg=green`:

nvim_search_current3

Which I think looks better. I wonder if we should change the default IncSearch settings?

(In order for the background to be green, I had to set the `ctermfg` (i.e, foreground), which seems weird.)

-------------

At this point, I think the basic prototype works exactly as expected. I have a few questions:

* How does nvim redraw the screen? When a redraw is asked, is the whole screen always redrawn from scratch, and then nvim compares to the old screen and figures out an efficient diff to be sent to the terminal (much like ncurses works)? Or does nvim have some elaborate update mechanism to only draw things that changed?
* Above I call `redraw_later(NOT_VALID)`, which I assume redraws the whole screen. Is it a waste to always redraw everything in `nv_next`?
* Should I figure out how to call the `redraw_buf_line_later` on the old line and the new line?

**Comment by certik** [comment]
This simplification can be applied:
```diff
diff --git a/src/nvim/buffer_defs.h b/src/nvim/buffer_defs.h
index dc0e718d6..16c7804be 100644
--- a/src/nvim/buffer_defs.h
+++ b/src/nvim/buffer_defs.h
@@ -940,7 +940,6 @@ typedef struct {
   buf_T       *buf;     // the buffer to search for a match
   linenr_T lnum;        // the line to search for a match
   int attr;             // attributes to be used for a match
-  int attr_cur2;        // attributes to be used for a current match
   int attr_cur;         // attributes currently active in win_line()
   linenr_T first_lnum;  // first lnum to search for multi-line pat
   colnr_T startcol;     // in win_line() points to char where HL starts
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index baa100333..766ad4657 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -3001,7 +3001,7 @@ win_line (
               if (wp->w_cursor.lnum == lnum
                   && wp->w_cursor.col >= (long)shl->startcol
                   && wp->w_cursor.col < (long)shl->endcol) {
-                shl->attr_cur = shl->attr_cur2;
+                shl->attr_cur = win_hl_attr(wp, HLF_I);
               } else {
                 shl->attr_cur = shl->attr;
               }
@@ -5613,7 +5613,6 @@ static void init_search_hl(win_T *wp)
   search_hl.lnum = 0;
   search_hl.first_lnum = 0;
   search_hl.attr = win_hl_attr(wp, HLF_L);
-  search_hl.attr_cur2 = win_hl_attr(wp, HLF_I);
 
   // time limit is set at the toplevel, for all windows
 }
```
and things still work. @justinmk should I do that? I don't know what style is preferred.

**Comment by justinmk** [comment]
> should I do that? I don't know what style is preferred

Sure.

Note that Bram seems to prefer the SearchCurrent approach, so we should probably stick with that. It can be linked with IncSearch by default.

**Comment by certik** [comment]
Sure, I'll revert it. It should probably be called something like `SearchLastMatch`, per Bram's comment.

**Comment by certik** [comment]
I updated the PR. Interesting is that before one did:
```
hi IncSearch ctermfg=green
```
but now one has to do:
```
hi SearchLastFoundMatch ctermbg=green ctermfg=black
```
to get the same result. It looks like the IncSearch is doing some inverting.

Anyway, the PR is now really simple. Just four very little changes.

@justinmk do you think that this change looks good? If so, I'll start working on updating the tests to pass.

**Comment by janlazo** [comment]
Cannot merge without tests.

**Comment by clason** [comment]
Bump -- that would be quite nice to have! @certik would you be interested in finishing this (rebase, docs, and test)?

@bfredl is this approach still fine (especially the forced redraw in `nv_next`)?

File: src/nvim/normal.c [hunk 1]
```diff
@@ -4890,6 +4890,10 @@ static void nv_next(cmdarg_T *cap)
     (void)normal_search(cap, 0, NULL, SEARCH_MARK | cap->arg, NULL);
     cap->count1 -= 1;
   }
+
+  // Redraw the buffer to ensure the highlighting of the current match under
+  // the cursor as well as the old match both get properly updated
+  redraw_later(NOT_VALID);
```
Comment (clason) [inline src/nvim/normal.c:4896]:
```suggestion
  redraw_later(curwin, NOT_VALID);
```

**Review by zeertzjq — COMMENTED** [review]
A change in `cursor_spec.lua` should be needed:
```diff
diff --git a/test/functional/ui/cursor_spec.lua b/test/functional/ui/cursor_spec.lua
index 03cd4bfd0..4c51547e2 100644
--- a/test/functional/ui/cursor_spec.lua
+++ b/test/functional/ui/cursor_spec.lua
@@ -212,10 +212,10 @@ describe('ui/cursor', function()
         if m.blinkwait then m.blinkwait = 700 end
       end
       if m.hl_id then
-          m.hl_id = 60
+          m.hl_id = 61
           m.attr = {background = Screen.colors.DarkGray}
       end
-      if m.id_lm then m.id_lm = 61 end
+      if m.id_lm then m.id_lm = 62 end
     end
 
     -- Assert the new expectation.
```

BTW the `SearchLastFoundMatch` name is really too long.

**Comment by numToStr** [comment]
I think setting the highlight group somehow affects the matchit highlighting

https://user-images.githubusercontent.com/24727447/147657618-cbc80387-e3ec-418e-922b-55f18da9250c.mp4

Note: I changed the hl-group name to `SearchCurrent`

**Comment by certik** [comment]
This would be very nice to have, but I need help from neovim developers to help me get this over the finish line. I need help with writing tests, as well as feedback if any changes are needed.

**Comment by clason** [comment]
Before any of that, we might want to bikeshed on the name; I agree that `SearchLastFoundMatch` is a mouthful and would prefer `SearchCurrentMatch` or even `SearchCurrent`. (I know Bram has issues with that name as it's not technically descriptive -- but that's what the documentation is for!)

---

Then, I think the first thing would be to add the new highlight group here:
https://github.com/neovim/neovim/blob/c46f7caad0bef587d189de2310435f48f9bd9ad8/src/nvim/option.c#L263
and here:
https://github.com/neovim/neovim/blob/c46f7caad0bef587d189de2310435f48f9bd9ad8/src/nvim/syntax.c#L6164
(probably easiest to link it to `IncSearch` by default -- or `Search`, in the hope that this would require fewer changes).

Documentation goes here: https://github.com/neovim/neovim/blob/c46f7caad0bef587d189de2310435f48f9bd9ad8/runtime/doc/syntax.txt#L5170-L5172

For tests, a good starting point would be to adapt one of the tests here: https://github.com/neovim/neovim/blob/master/test/functional/ui/searchhl_spec.lua
(I'm not a test expert, so I hope someone else will step in here, as that can indeed be rough the first time ;))

**Comment by famiu** [comment]
Do you mind if I take over this PR? I think I might be able to finish it if you don't have time

**Comment by clason** [comment]
@famiu I think you can take the silence for a "yes".

**Comment by famiu** [comment]
> @famiu I think you can take the silence for a "yes".

Will work on it once I get the PC

**Comment by leungbk** [comment]
An initial attempt at this has been merged upstream (https://github.com/vim/vim/pull/10133#issuecomment-1094121087), with some small issues.

**Comment by certik** [comment]
@famiu yes, please take over the PR. I don't have the time to learn enough about neovim to bring it over the finish line, it looks like tests have to be written etc.

**Comment by zeertzjq** [comment]
Superseded by #18081

**Pull request closed without merging**